Libraries/Image/RCTImageLoader.m

```diff
@@ -480,6 +470,18 @@ - (RCTImageLoaderCancellationBlock)loadImageWithURLRequest:(NSURLRequest *)image
   __block void(^cancelLoad)(void) = nil;
   __weak RCTImageLoader *weakSelf = self;
 
+  // Load handlers are executed before the generic caching logic so they are responsible
+  // for caching the images if necessary.
+  id<RCTImageURLLoader> loadHandler = [self imageURLLoaderForURL:imageURLRequest.URL];
```

Comment:
I'd rather expose an (optional) API on the loadHandler asking for things like `shouldCacheImageForURL`? Then if you want to, you can skip that part. That doesn't solve the threading approach though.